extensions/GitExtension/FileExplorerGitIntegration/Models/GitCommandRunner.cs

```diff
@@ -0,0 +1,229 @@
+
+    private readonly ILogger log = Log.ForContext<GitCommandRunner>();
+
+    public GitCommandRunner(string? path)
+    {
+        if (RuntimeHelper.IsMSIX)
+        {
+            GitExecutableConfigOptions = new GitExecutableConfigOptions
+            {
+                GitExecutableConfigFolderPath = ApplicationData.Current.LocalFolder.Path,
+            };
+        }
+        else
+        {
+            GitExecutableConfigOptions = new GitExecutableConfigOptions
+            {
+                GitExecutableConfigFolderPath = path ?? string.Empty,
+            };
+        }
+
+        fileService = new FileService();
+        ReadInstallPath();
+
+        fileWatcher = new FileSystemWatcher(GitExecutableConfigOptions.GitExecutableConfigFolderPath, GitExecutableConfigOptions.GitExecutableConfigFileName);
+        fileWatcher.NotifyFilter = NotifyFilters.LastWrite;
+        fileWatcher.Changed += OnFileChanged;
+        fileWatcher.EnableRaisingEvents = true;
+        log.Debug("FileSystemWatcher initialized for configuration file");
+    }
+
+    private void OnFileChanged(object sender, FileSystemEventArgs args)
+    {
+        if (args.Name == GitExecutableConfigOptions.GitExecutableConfigFileName)
+        {
+            ReadInstallPath();
+        }
+    }
+
+    public void ReadInstallPath()
+    {
+        GitExeInstallPath = fileService.Read<string>(GitExecutableConfigOptions.GitExecutableConfigFolderPath, GitExecutableConfigOptions.GitExecutableConfigFileName);
+    }
+
+    public bool IsGitExeInstallPathSet()
+    {
+        return !string.IsNullOrEmpty(GitExeInstallPath);
+    }
+
+    public GitCommandRunnerResultInfo ValidateGitExeInstallPath()
+    {
+        try
+        {
+            System.Diagnostics.Process.Start(GitExeInstallPath);
+            return new GitCommandRunnerResultInfo(ProviderOperationStatus.Success, null);
+        }
+        catch (Exception ex)
+        {
+            log.Error(ex, "Failed to start Git.exe at configured path");
+            return new GitCommandRunnerResultInfo(ProviderOperationStatus.Failure, "Failed to start Git.exe at configured path", string.Empty, ex, null);
+        }
+    }
+
+    public bool StoreGitExeInstallPath(string path)
+    {
+        log.Information("Setting Git Exe Install Path");
+        GitExeInstallPath = path;
+        fileService.Save(GitExecutableConfigOptions.GitExecutableConfigFolderPath, GitExecutableConfigOptions.GitExecutableConfigFileName, GitExeInstallPath);
+        log.Information("Git Exe Install Path stored successfully");
+        return true;
+    }
+
+    public bool DetectGit()
+    {
+        var gitExeFound = false;
+
+        if (!gitExeFound)
+        {
+            // Check if git.exe is present in PATH environment variable
+            try
+            {
+                System.Diagnostics.Process.Start("git.exe");
+                gitExeFound = true;
+                StoreGitExeInstallPath("git.exe");
```

Comment:
why not run "where git" in cmd and parse the output?
C:\>where git
C:\Program Files\Git\cmd\git.exe
D:\work\depot_tools\git.bat

Reply:
Mostly we're trying to be somewhat resilient to users not having git.exe on the path, and try some sensible fallback directories. But to be fair, the first option we try is the one the path.

Or am I misremembering that "where" depends on the PATH variable?

Reply:
you're correct where depends on the path variable. my comment was why execute "git.exe" to figure out if its on the path when we have an existing utility that will tell us if it is and precisely in what location.

Reply:
I think we're sort of agreeing with each other. From the point of this code, if it's on the PATH, we don't care where it is - we'll just execute git.exe without qualifying the path. The other options are for "Hmm... it's not on PATH. Let's try a few other well-known locations to see if we can find something that works".